src/StepperISR_due.cpp

```diff
@@ -1,401 +1,691 @@
+          e->ticks;  // Change the period first!
+      PWM_INTERFACE->PWM_IDR1 = mask;
+      uint32_t dwSR = port->PIO_ABSR;
+      port->PIO_ABSR = (g_APinDescription[mapping->pin].ulPin | dwSR);
+      // Remove the pins from under the control of PIO
+      port->PIO_PDR = g_APinDescription[mapping->pin].ulPin;
+      port->PIO_IER = g_APinDescription[mapping->pin].ulPin;
+      q->_pauseCommanded = false;
+    } else {
+      // We needed to pause longer, so set the period, and wait...
+      // set the period, reconnect, restart, then disconnect PWM interrupt, so
+      // the next interrupt goes to the PIO interrupt handler and works like
+      // normal
+      PWM_INTERFACE->PWM_CH_NUM[channel].PWM_CPRDUPD = e->ticks;
+    }
+    q->read_idx = rp;
+  }
+}
+
+inline void disconnectPWMPeriphal(Pio* port, uint8_t pin,
+                                  uint32_t channelMask) {
+  // This function has become quite minimal.  Its still somewhat nice to have
+  // it seperated out, but I may decided it needs to just be inline...2 lines
+  // saved and function call overhead added :-/
+  PWM_INTERFACE->PWM_DIS = channelMask;
+  PIO_SetOutput(port, g_APinDescription[pin].ulPin, 0, 0, 0);
+}
+
+inline void attachPWMPeripheral(Pio* port, uint8_t pin, uint32_t channelMask,
+                                bool clearISR = false) {
+  PWM_INTERFACE->PWM_IDR1 = (channelMask);
+  uint32_t dwSR = port->PIO_ABSR;
+  port->PIO_ABSR = (g_APinDescription[pin].ulPin | dwSR);
+  // Remove the pins from under the control of PIO
+  port->PIO_ODR |= g_APinDescription[pin].ulPin;
+  port->PIO_PDR |= g_APinDescription[pin].ulPin;
+  port->PIO_IER |= g_APinDescription[pin].ulPin;
+  PWM_INTERFACE->PWM_ENA |= channelMask;
+}
+#define ISRQueueName(Q) Queue##Q##ISR
+#define ISRQueueNameStr(Q) "Queue" #Q "ISR"
+
+#define DUE_STEPPER_ISR(Q)                                                     \
+  void Queue##Q##ISR() {                                                       \
+    /*These need only be looked up once rather than multiple times in the      \
+     * isr*/                                                                   \
+    /*One would hope the compiler would do this, but how is it supposed to*/   \
+    /*know if these values change or not?  const requires setting it at */     \
+    /*instantiation or casting...this is the next best thing :) */             \
+    static StepperQueue* const q = &fas_queue[Q];                              \
+    static uint32_t channel = fas_queue[Q].mapping->channel;                   \
+    static uint32_t samPin =                                                   \
+        g_APinDescription[fas_queue[Q].mapping->pin].ulPin;                    \
+    static const PWMCHANNELMAPPING* mapping = fas_queue[Q].mapping;            \
+    static Pio* port = fas_queue[Q].mapping->port;                             \
+    if (!q->_hasISRactive || q->_pauseCommanded) {                             \
```

Comment:
In the above commands, `fas_queue[Q].` can be replaced with `q->`.
Should be faster, if the compiler cannot see this optimization.

Reply:
hmmm, its static, so I thought it was initialized only once.  If the compiler isn't doing that, then sure, removing an indirection helps....however, there are other optimizations that could be made....I didn't go right for them and to them because I'm not sure how big an impact they will have, and technically, if the optimizer is doing its job (something I don't have a lot of faith in having hand optimized X86/SSE and AMDIL GPU code) it *should* be recognizing constants looked up multiple times like mapping->channel and storing them in a register that persists throughout the function...I have seen where the compiler is adamant about not storing it in a register even when register pressure is -20!  So, perhaps a static const would be warranted for many of these variables.  However, I had working code, I wanted it committed before I went off making changes, and further, this was a distraction from my main project using the Due, amd so while I don't want to forget about this, working code is working code, which means I can move on to the mower, get that working, then in the winter, when inevitable downtime shows up, I can come back and see the assembly the compiler is generating and decide if its worth the changes in C code.  I know one side likes to argue that the compiler can see optimizations like this and is good at it, but I've seen too many times where it just doesn't, so I feel its worth the check.  (For example:  Get Visual Studio, GCC, Clang, or whatever your favorite optimizing compiler is to use RSI/RDI for accumulation.  Even when you do not need another pointer for the rest of the function.  It simply will not, and will throw stuff on the stack!)

Reply:
good point. I have not really seen, that this is defined as static per channel. It just looks bit inconsistent, if a pointer is defined, but is not always used.

And yes, first push a working code and polish later is same, what I do. And this comment should only support the polishing.